### ISSUE TITLE
LogitsPostProcessor BatchManager bug with tp > 1

### ISSUE THREAD

**Issue by akhoroshev** [NVIDIA/TensorRT-LLM#1415]
Generation breaks if I modify logits (with [setLogitsPostProcessor](https://github.com/NVIDIA/TensorRT-LLM/blob/main/cpp/include/tensorrt_llm/batch_manager/inferenceRequest.h#L157)) with tp 2. GptManager freezes on first iteration. I call setLogitsPostProcessor only for rank 0 node (I do it in this [function](https://github.com/NVIDIA/TensorRT-LLM/blob/main/benchmarks/cpp/gptManagerBenchmark.cpp#L877C35-L877C46)).

But it can be fixed if i call setLogitsPostProcessor for all nodes (for example i can call setLogitsPostProcessor [there](https://github.com/NVIDIA/TensorRT-LLM/blob/main/benchmarks/cpp/gptManagerBenchmark.cpp#L806) for all ranks).

Why I need to set callback for all nodes? Is this a bug or bad documentation?

**Comment by byshiue** [comment]
Tensor Parallelism is not applied on logit Processor. Namely, every rank gets same logits and do sam logit processor at the same time. So, if you change one of them, this rank would have different results and lead to dead lock. So, you must change all ranks.

**Comment by akhoroshev** [comment]
We found the best solution: using nccl bcast across all ranks
